### ISSUE TITLE
ChainWatchers.notify errors on _notifyProperties by ED

### ISSUE THREAD

**Issue by knownasilya** [emberjs/ember.js#13470]
``` no-highlight
TypeError: Cannot read property 'notify' of undefined
    at Object.ChainWatchers.notify (http://crcogct.localhost:3000/dashboard/assets/vendor.js:28475:17)
    at chainsDidChange (http://crcogct.localhost:3000/dashboard/assets/vendor.js:34505:9)
    at Object.propertyDidChange (http://crcogct.localhost:3000/dashboard/assets/vendor.js:34415:5)
    at contentPropertyDidChange (http://crcogct.localhost:3000/dashboard/assets/vendor.js:46339:32)
    at Object.apply (http://crcogct.localhost:3000/dashboard/assets/vendor.js:37154:18)
    at Object.sendEvent (http://crcogct.localhost:3000/dashboard/assets/vendor.js:30699:28)
    at ObserverSet.flush (http://crcogct.localhost:3000/dashboard/assets/vendor.js:34022:25)
    at Object.endPropertyChanges (http://crcogct.localhost:3000/dashboard/assets/vendor.js:34533:19)
    at _notifyProperties (http://crcogct.localhost:3000/dashboard/assets/vendor.js:127142:25)
    at InternalModel.adapterDidCommit (http://crcogct.localhost:3000/dashboard/assets/vendor.js:126466:19)
```

I'm using Ember 2.5.1 and Ember Data 2.5.2.

The error seems like it could be avoided on the Ember side. Not sure what would make the node undefined, but there should probably be a check, because otherwise it rejects the `model.save()` promise (although the save works, but any `then`s do not fire).

> Original issue https://github.com/lytics/ember-data-model-fragments/issues/198

**Comment by knownasilya** [comment]
@krisselden I see you are one of the people who contributed there. Any thoughts?

**Comment by acorncom** [comment]
This one looks like it'll be fun to try to reproduce ;-)

**Comment by krisselden** [comment]
This happens when a path is unwatched that was never watched.  Usually something is manually trying to teardown an observer that was never set up.

**Comment by krisselden** [comment]
unfortunately this is unlikely to get much attention without more information.

**Comment by richpwf** [comment]
I've run into this, I can replicate it in my app but the circumstances that it happens in are rather unique. It seems to be tied to using removeObject on Embedded Records in my case but only if I enter the parent route first (if I got straight to the child route then I don't get the error). 

Happy to try recording a video or connecting with someone wiser than me to help debug as it's forced me to remove functionality at the moment.

**Comment by knownasilya** [comment]
@richpwf think you could replicate it?

**Comment by richpwf** [comment]
@knownasilya sadly not, I tried at the time but couldn't get it to happen in a twiddle (my actual app is pretty large, ~1000 files, so replicating all of it isn't an option)

**Comment by knownasilya** [comment]
Seems like it happens when Ember tries to notify the `[]` e.g. `items.[]` key.

**Comment by knownasilya** [comment]
@krisselden how could something teardown an observer? What would I look for?

**Comment by krisselden** [comment]
Generally this comes from something always running teardown in a code path where setup doesn't always happen.

**Comment by krisselden** [comment]
Which is why it is next to impossible for us to investigate without more info as to where the imbalance is coming from.

**Comment by knownasilya** [comment]
@krisselden I can give you access to my app, if that would help. I'm just not sure what I'm debugging..

**Comment by knownasilya** [comment]
So I think I've nailed it down, but having a hard time reproducing.

Seems like it only happens if have a certain computed array is in the template..

```js
savedThemes: computed('community.settings.map.thematicMaps.length', function () {
    var themes = this.get('community.settings.map.thematicMaps');
    return themes.filterBy('hasDirtyAttributes', false);
  }),
```

```hbs
<div class="row">
  <div class="col-xs-12">
    {{#if savedThemes}}
      {{#each community.settings.map.thematicMaps as |item|}}
        {{theme-item theme=item class='theme-item' togglepublish=(route-action 'saveCommunity')
          edit=(route-action 'editTheme')}}
      {{/each}}
    {{else}}
      <p class="lead">
        No Themes, create one {{#link-to 'map.themes.new'}}here{{/link-to}}.
      </p>
    {{/if}}
  </div>
</div>
```

If I remove the `savedThemes` array (if I even have `{{log savedThemes}}` it errors), it works.

**Comment by knownasilya** [comment]
@krisselden Ok, I have something. The error is different but it's caused the same way (CP present).

https://github.com/knownasilya/notify-error

To get the error to appear click 'Approve' on the first comment.
If you remove `approvedComments` from `app/templates/application.hbs` the error goes away.

I think it has to do with the fact that I'm using an ObjectProxyPromise as my service..

**Comment by krisselden** [comment]
@pixelhandler the issue in the reproduction is an infinite loop,not the above error and seems to be an issue with model fragments and ED. This hasn't been demonstrated to be an ember bug

**Comment by christophersansone** [comment]
I am coming from https://github.com/lytics/ember-data-model-fragments/issues/173 , which is the `1.13.13` version of this error.  I can consistently reproduce it in an app I'm building, and I have been able to set a conditional breakpoint to step through the process immediately before the error occurs.  I'd be happy to to a screen share with @krisselden or anyone else who may have some input.  I'm not convinced I can create a minimum test case, but this may be the next best thing.

I'll try to explain it here.  The problem happens in `chainsDidChange`:

```javascript
for (i = 0, l = nodes.length; i < l; i++) {
  nodes[i].didChange(events);
}
```

In my case, `nodes.length` is initially 3.  Once the first iteration of this loop occurs, `nodes.length` becomes 1.  Therefore, `nodes[i]` is out of bounds for the second iteration.  Why does that happen?  Let's dig in deeper.

In the top-level call of `chainsDidChange`, `keyName` is a computed property called `object`, which references a model fragment (`MF.fragment`).  So `nodes[i]` is the ChainNode for the `object` property.  Within `ChainNode.didChange()`, it then calls `didChange` on the watched properties of `object`.  The first property is called `top`, so it calls `didChange` for that key.  Now here's the kicker...

Within `didChange` for `top`, it gets its parent (i.e. `object`) via:

```javascript
var obj = this._parent.value();
```

Prior to this call, `this.parent._value` is undefined.  Within the call to `value()`, it evaluates the `object` computed property.  The evaluation of `object` is what is causing the content of `nodes` to change.  Prior to `value()`, `nodes.length` is 3.  After `value()`, `nodes.length` is 1.

*Long story short: when an object notifies its observers of changes, the object is recomputed which changes the list of observers.*

A few questions that may help us get to the bottom of it:

1.  Is it unexpected behavior for chain watchers to change due to a CP evaluation?  I would assume that the observers are determined ahead of time, and a CP evaluation would not affect that.  It seems that `chainsDidChange` is not expecting the list to change.

2.  What would cause the chain watchers to change?  Anything other than explicit calls to `addObserver` / `removeObserver` outside of the Ember internals?

Thanks, hope this helps, and any help anyone else can give is greatly appreciated!

**Comment by christophersansone** [comment]
Quick follow-up.  I changed the `chainsDidChange` loop from:

```javascript
for (i = 0, l = nodes.length; i < l; i++) {
  nodes[i].didChange(events);
}
```

to:

```javascript
for (i = 0; i < nodes.length; i++) {
  nodes[i].didChange(events);
}
```

And the issue no longer occurs.  This may have other implications though.  For any app that does not experience this problem, `nodes.length` will always equal `l`, so I cannot see a case where it would cause a defect.  The constant evaluation of `nodes.length` may be a performance regression, but I would like to think that modern Javascript engines have optimized for this and cache the value when the array has not changed.

I'm happy to submit a PR for this.  I imagine it may be covering up an unexpected implementation issue in `ember-data-model-fragments`, and I'm happy to help diagnose that some more as well.  Even if we resolve the issue in `ember-data-model-fragments`, maybe at the very least this method can throw a more descriptive error once we understand the root cause.

**Comment by christophersansone** [comment]
It looks like #11564 was a similar issue a year ago in a different file.  The fix was similar to what I suggested:

```javascript
for (var i = 0, l = chainNodes.length; i < l; i++) {
  var node = chainNodes[i];
  if (node) {
    node.didChange(null);
  }
}
```

And according to the conversation, a couple awesome core team members agreed that both `didChange` and `willChange` should be guarded similarly.

PR for 1.x is on its way, though not sure that 1.x is maintained anymore.

@knownasilya would you be up for submitting a similar PR for 2.x?

**Comment by knownasilya** [comment]
Sure, I'll have a look this weekend

**Comment by mmun** [comment]
I'm closing this issue as it's not clear that it is an error with Ember. Please chime in if you'd like us to reopen this issue.
